08.ls_object/lib/ls.rb

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'etc'
+require 'optparse'
+require 'io/console/size'
+
+require_relative 'filedata'
+require_relative 'filelist'
+require_relative 'display'
+
+class Ls
+  attr_reader :display
+
+  def call
+    options = parse_options
+    target = ARGV[0] || '.'
+
+    return print "ls: #{target} : No such file or directory" unless File.exist?(target)
```

Comment:
コードがすっきりしたら別のことが気になりました。ターゲットがなければ、これはうまく動かないっていうことだから、エラー扱いでよさそう。

```suggestion
    raise "ls: #{target} : No such file or directory" unless File.exist?(target)
```

こちらのほうが意図に近いのでは?

Reply:
7310e20

意図としては確かに例外を出したほうがいいのかなと思う一方で、プラクティスとしては本家の見た目に近づけるというのが合ったと思うので、`return print~`としていました。

`raise`を返したときの表示でなるべく本家に近い形を目指して、現状は以下の表示になっています。

```zsh
❯❯❯ % ruby lib/ls.rb -l hoge  
lib/ls.rb:20:in `ls': hoge is no such file or directory (ArgumentError)
	from lib/ls.rb:52:in `<main>'
```

こっちは本家
```zsh
❯❯❯ % ls -la hoge 
ls: hoge: No such file or directory
```

Reply:
課題の意図からすると例外が抜けていっても良いと思いますが、「本家に近い形」を目指すのだととすると、`print`で抜けると正常終了なので行儀はよくなさそうです。

ちゃんと試してませんが呼び出す側から、

```ruby
begin
  puts Ls.new.ls
rescue ArgumentError => e
  puts e.message
  exit 1
end
```
とすると、`ls`の挙動に近くなりそうです。

Reply:
dfc9a91

![image](https://user-images.githubusercontent.com/5976902/108223837-959ee000-717d-11eb-8646-b04a2ac54b5a.png)

✨

Reply:
豆知識です。

Unixプロセスには終了ステータスがあり([終了ステータス \- Wikipedia](https://ja.wikipedia.org/wiki/%E7%B5%82%E4%BA%86%E3%82%B9%E3%83%86%E3%83%BC%E3%82%BF%E3%82%B9))、`ls`も `ls -a hoge` としたら`1`を返します(zshなら `$?`に入ってます)。
```
$ls hoge; echo $?
ls: hoge: No such file or directory
1
```
正常終了は`0`です。